execution_engine/src/core/runtime_context/mod.rs

```diff
@@ -704,14 +746,23 @@ where
     /// values can't be added, either because they're not a Monoid or if the
     /// value stored under `key` has different type, then `TypeMismatch`
     /// errors is returned.
-    pub fn add_gs(&mut self, key: Key, value: StoredValue) -> Result<(), Error> {
+    pub(crate) fn metered_add_gs<K, V>(&mut self, key: K, value: V) -> Result<(), Error>
+    where
+        K: Into<Key>,
+        V: Into<StoredValue>,
+    {
+        let key = key.into();
+        let value = value.into();
         self.validate_addable(&key)?;
         self.validate_key(&key)?;
         self.validate_value(&value)?;
-        self.add_unsafe(key, value)
+        self.metered_add_gs_unsafe(key, value)
     }
 
-    fn add_unsafe(&mut self, key: Key, value: StoredValue) -> Result<(), Error> {
+    fn metered_add_gs_unsafe(&mut self, key: Key, value: StoredValue) -> Result<(), Error> {
```

Comment:
Shouldn't `ADD` be more expensive than `WRITE` since it requires first reading the value before writing it back to the GS?

Reply:
@goral09 you always have to pay for `new_uref` first, so I'd say yes - add is more expensive even though we don't charge for reads. We don't plan to charge for reads alone though CC @AlexanderLimonov

Reply:
I don't understand where paying for `new_uref` comes from.

Reply:
Depends on what we mean by "reads"

Reply:
Reads from the global state I suppose. But we don't have plans to charge for it, do we? We can take this discussion offline @goral09 @AlexanderLimonov

Reply:
If it takes time during deploy execution phase, it should have a gas cost.

Reply:
Egzactly! :grimacing: